modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/NonAtomicFields.scala

```diff
@@ -47,10 +48,21 @@ object NonAtomicFields {
 
   def resolveTypes[F[_]: Sync: RegistryLookup](
     resolver: Resolver[F],
-    entities: Map[TabledEntity, Set[SchemaSubVersion]]
+    entities: Map[TabledEntity, Set[SchemaSubVersion]],
+    filterCriteria: List[SchemaCriterion]
   ): F[Result] =
     entities.toList
+      .map { case (tabledEntity, subVersions) =>
+        tabledEntity -> filterSubVersions(filterCriteria, tabledEntity, subVersions)
+      }
+      .filter { case (_, subVersions) =>
+        // Remove whole schema family if there is no subversion left after filtering
+        subVersions.nonEmpty
+      }
       .traverse { case (tabledEntity, subVersions) =>
+        // Even though we may filter some of the subVersions earlier, we still fetch all available from Iglu (same family ~> same model).
+        // It's because we can't have 'holes' in schema family when building typed entities. Otherwise we may end up with invalid and incompatible merged schema model.
+        // As a result, it's only possible to filter out whole schema family, but not part of the family.
```

Comment:
> As a result, it's only possible to filter out whole schema family

That's... kind of true.  But this filtering method does let us filter out certain sub-versions from what gets loaded to the warehouse.

The `Result` returned by `resolveTypes` now contains a filtered list of sub-versions in the `TypedTableEntities`.  So later, when we call [`Transform.transformEvent`](https://github.com/snowplow-incubator/common-streams/blob/0.4.0/modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/Transform.scala#L43-L57), we only pick out non-filtered entites [on this line](https://github.com/snowplow-incubator/common-streams/blob/0.4.0/modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/Transform.scala#L222-L224).

Reply:
I think... it's actually true. 

Looking at how `TypedTableEntity` is built [here](https://github.com/snowplow-incubator/common-streams/blob/50cddf9e16f58d8e0fcb65af6101f795495bf0d0/modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/TypedTabledEntity.scala#L56), `mergedVersions` field is not the one we filtered before. It's built using schemas fetched from Iglu, which are not filtered.

Reply:
Oh yeah you're correct.

So we _could_ amend [this line](https://github.com/snowplow-incubator/common-streams/blob/50cddf9e16f58d8e0fcb65af6101f795495bf0d0/modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/TypedTabledEntity.scala#L81) so we only conditionally add the subversion to the set.

The way you've done it... I think it is still possible to filter out the highest subversion in a family.  For example, if the trackers are sending `1-0-0` and `1-0-1`, and if you set `skipSchemas: 1-0-1` then I think the `TypedTabledEntity` includes only `1-0-0`.

So the way you've done it is still quite helpful for many purposes.  Do you think it's worth also making that change to `TypedTabledEntity.build`?

Reply:
>  I think it is still possible to filter out the highest subversion in a family

Yes, that's correct. Looks like it's because of this [line](https://github.com/snowplow-incubator/common-streams/blob/50cddf9e16f58d8e0fcb65af6101f795495bf0d0/modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/SchemaProvider.scala#L43) where we have another filter in `SchemaProvider`

> Do you think it's worth also making that change to `TypedTabledEntity.build`

I decided to implement this, not in `TypedTabledEntity.build`  but in `NonAtomicFields.resolveTypes` to keep everything about filtering in the same place https://github.com/snowplow-incubator/common-streams/pull/61/commits/7300d5363b4a7e844fbffe219649c43b43abdf5c + comments adjustments